prov/rxm/src/rxm_rma.c

```diff
@@ -71,32 +307,48 @@ rxm_ep_rma_common(struct rxm_ep *rxm_ep, const struct fi_msg_rma *msg, uint64_t
 	struct fi_msg_rma msg_rma = *msg;
 	struct util_cmap_handle *handle;
 	struct rxm_conn *rxm_conn;
-	void *mr[RXM_IOV_LIMIT] = { 0 };
+	void *mr_desc[RXM_IOV_LIMIT] = { 0 };
 	int ret;
 
 	ret = ofi_cmap_get_handle(rxm_ep->util_ep.cmap, msg->addr, &handle);
-	if (OFI_UNLIKELY(ret))
-		return ret;
+	if (OFI_UNLIKELY(ret)) {
+		if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+			return ret;
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		if (handle->state != CMAP_CONNECTED) {
```

Comment:
This flow looks odd.  ofi_cmap_get_handle() fails, but we still expect handle to be valid.

Reply:
Yes, I agree. But we should handle case here that connection is successfully established in CM thread when trying to hold `cmap::lock` lock from master thread.
The another way is to remove acquiring of `cmap::lock` lock and each calling of `ofi_cmap_get_handle` has to hold `cmap::lock`... Not sure that this is right way

Reply:
I think the best option for this code flow (and any that are similar) is to separate ofi_cmap_get_handle() into two calls -- cmap_acquire_handle() and handle_connect() -- but with better names.
```
/* top half of ofi_cmap_get_handle */
ofi_cmap_acquire_handle(...)
{
    handle = util_cmap_get_handle(...)
    if (!handle)
        util_cmap_alloc_handle(..., &handle);
    return handle
}

/* bottom half of ofi_cmap_get_handle */
ofi_handle_connect(...)
{
    if (handle->state == CMAP_CONNECT)
        return 0;
    switch (handle->state) {
        ...
    }
    return -FI_EAGAIN;
}
```
This code can then be restructured to avoid the double lock acquire-release flow.  E.g.
```
lock(...)
handle = ofi_cmap_acquire_handle(...);
if (handle && handle->state != CMAP_CONNECTED) {
    ...
}
```
This needs more work, but that's the general idea.

Reply:
I should have added that ofi_cmap_get_handle() would be modified to call these 2 new functions, in case that wasn't obvious.

Reply:
We could also add a callback to ofi_cmap_get_handle which would be called if the connection is not established yet. The callback can then add the send to postponed list. That would keep the lock-release code confined to cmap.

Reply:
Thanks! This makes sense.
I've created separate PR #3879 that addresses this comment for send/inject calls

@a-ilango it would be great to have this, but unfortunately we should build arg to pass them into `ofi_cmap_get_handle()` before calling this. This adds some overhead